### DISCUSSION TITLE
The weirdest thing I've been seen in years

### DISCUSSION THREAD

**Discussion by ruifig** [earlephilhower/arduino-pico#1381] [General]
So, I encountered this mind boggling problem with my complex IoT project, but just to make sure I'm not going crazy, I created a minimal sample using (https://github.com/maxgerhardt/platform-raspberrypi#5a533d6b36844183bb72794bc53091206165e921 ), and I'm getting the same problem:
Sharing this here since I'm really running out of ideas on how the heck this is possible.
I have over 20 years experience with C++ (I'm a software engineer) and never seen a problem like this.

Very-very likely not related to Arduino-Pico, but I'm out of ideas, and maybe someone has seen something similar.

Scenario:

* Using PlatformIO
* Raspberry Pi Pico W
* Platform set to: https://github.com/maxgerhardt/platform-raspberrypi#5a533d6b36844183bb72794bc53091206165e921

Given this simple program:

```
#include <Arduino.h>

#if 0
  //#include "FreeRTOS.h"
#endif

void setup() {
  Serial.begin(115200);
  Serial.print("Hello world!");
}

void loop() {
  Serial.println(millis());
  delay(1000);
}
```
It compiles to this:
```
RAM:   [===       ]  26.4% (used 69244 bytes from 262144 bytes)
Flash: [          ]   4.4% (used 92492 bytes from 2093056 bytes)
```
and it works fine.

Now changing JUST the `#if 0` to this:
```
#if 0
  #include "FreeRTOS.h"
#endif
```
It compiles to this:
```
RAM:   [===       ]  29.5% (used 77416 bytes from 262144 bytes)
Flash: [=         ]   5.2% (used 109416 bytes from 2093056 bytes)
```
And the firmware doesn't work.

* First, how the heck is something inside a `#if 0` affecting my build (the firmware size clearly shows that). Note that there is no other code. I understand that MAYBE this could be pranked by doing a `#define 0 1` or something like that, but that this is a minimal sample I created just now.
* Second, the firmware stops working. I can understand that maybe this could happen if by whatever magic is actually adding "FreeRTOS.h" to my build, maybe "FreeRTOS.h" introduces something that breaks the firmware, but then for the kicks I did this:

```
#if 0
  #error PEEKABOOO
  #include "FreeRTOS.h"
#endif
```

And it still builds, clearly showing that `#error` is not seen.

Anyone has encountered this before? haha

**Comment by earlephilhower** [comment]
That's the Arduino IDE doing it's thing.  It doesn't understand `#if` so when it scans the INO (to do library dependencies and add prototypes behind the scenes to make the INO a regular CPP file) it automatically links in the FreeRTOS code in the failing case.  The rest of the core, however, isn't aware of it and things go weird.

Moral of the story is to use comments not if blocks to remove includes in the main INO.  For other .c files, I don't believe the IDE scans then so you can use if blocks there, if you really want.

**Reply by ruifig** [reply]
Thanks for insight about how the Arduino IDE figures out dependencies, but some more details:

* I'm using PlatformIO, not the Arduino IDE
* This is a main.cpp file, not a INO file, in case it makes a difference to those Arduino IDE shenanigans.

**Reply by ruifig** [reply]
Ok, my guess is that PlatformIO still ends up using the Arduino IDE somehow, or uses a similar process to help/suggest dependencies?

**Reply by earlephilhower** [reply]
I would imagine so.  You can check by looking at the MAP files and seeing if there are FreeRTOS functions present.  In any case, it is weird if you're not expecting it!

**Reply by ruifig** [reply]
I already checked yesterday through PlatformIO's "Inspect" menu, and the extra size is indeed coming from FreeRTOS symbols, so that's what is happening as you suggested.
This was unexpected to me because I wasn't aware the Arduino IDE (or PlatformIO as it seems) would do such a thing behind the scenes.
Plenty of experience with consoles and desktop (that's what I do), but electronics is just a hobby for me and I'm learning these tools and quirks as I go, reading a minimum of documentation. ;)
Anyway, I guess such approach makes things easier for newcomers to help with dependencies. But for someone who is expecting the code to do what it should and no more, this just gets in the way.

Now that I know this is a thing, there is indeed plenty of of discussions I can find online, such as ( https://github.com/arduino/Arduino/issues/1841 ) . 👍 

Again, thanks for the help, Earle.

**Comment by ruifig** [comment]
And of course, now that I know what's going on, I can see PlatformIO has settings to change this behaviour ( [https://docs.platformio.org/en/latest/librarymanager/ldf.html#ldf-mode](url) 

The default is `chain` which doesn't evaluate preprocessor conditional syntax.

**Reply by maxgerhardt** [reply]
Was about to tell you this. It blindly scans the `#include` line in the default mode which doesn't respect `#if` conditions. This routinely throws off people that are using PlatformIO and is subject to improvment in e.g. https://github.com/platformio/platformio-core/issues/3113

**Comment by vogon** [comment]
hey, I just ran into this issue two years later but couldn't find the discussion until I was about to post my own, so I'm going to leave more notes here so hopefully more people stumble across it.  the LVGL Arduino library creates a conditional dependency on FreeRTOS that, similarly, automatically links in FreeRTOS even if `LV_USE_OS` is set to `LV_OS_NONE`, and even if you don't include `lvgl.h` or write a single line of LVGL code, and the fix is the same: switch `lib_ldf_mode` to `chain+` or another mode that is capable of identifying that the conditional dependency doesn't exist.
